### DISCUSSION TITLE
Decimate 3D model and maintain approximate point color.

### DISCUSSION THREAD

**Discussion by caleb-oconnor** [pyvista/pyvista#6284] [Q&A]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
I have a very large 3D mesh ~1.5 millions points and each point contains a color. If possible I would like to decimate the mesh while also retaining an approximated color for that point.

mesh = pv.PolyData(vertices, faces)
mesh['colors'] = vertices_color

ds_mesh = mesh.decimate(1 - (50000 / len(mesh.points)))
ds_mesh['colors']
```

### Description

KeyError: 'Data array (colors) not present in this dataset.'

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Fri Jun 21 13:06:48 2024 Central Daylight Time
                OS : Windows
            CPU(s) : 40
           Machine : AMD64
      Architecture : 64bit
               RAM : 63.7 GiB
       Environment : IPython
        GPU Vendor : NVIDIA Corporation
      GPU Renderer : Quadro RTX 8000/PCIe/SSE2
       GPU Version : 4.5.0 NVIDIA 471.11
  MathText Support : False
  Python 3.8.7 (tags/v3.8.7:6503f05, Dec 21 2020, 17:59:51) [MSC v.1928 64 bit
  (AMD64)]
           pyvista : 0.43.10
               vtk : 9.3.0
             numpy : 1.24.4
        matplotlib : 3.7.4
            scooby : 0.9.2
             pooch : 1.8.0
            pillow : 10.2.0
           imageio : 2.34.1
         pyvistaqt : 0.11.0
           IPython : 8.12.3
        ipywidgets : 8.1.1
             scipy : 1.10.1
      nest_asyncio : 1.6.0
--------------------------------------------------------------------------------
```

**Comment by darikg** [comment]
I think you should be able to do something like `ds_mesh = ds_mesh.sample(mesh)` to sample point data from the original mesh

**Comment by caleb-oconnor** [comment]
Thanks I have never seen this function. Unfortunately, I have been unable to get it to reproduce similar results. I have tried changing the tolerance parameter since I have reduced a lot of points but it gives a very blurry type result.
Original:
![Original](https://github.com/pyvista/pyvista/assets/93679057/906998d5-7169-4d82-a593-167851aa0fc6)
ds_mesh = ds_mesh.sample(mesh)
![Sample](https://github.com/pyvista/pyvista/assets/93679057/89ec2904-8c1e-4fd3-adda-78de862e6b54)
ds_mesh = ds_mesh.sample(mesh, tolerance=0.01)
![Sample_Tolerance_0 01](https://github.com/pyvista/pyvista/assets/93679057/c76ebc7c-ca5e-423a-8ec9-924254533cd7)

**Comment by darikg** [comment]
Hmm maybe `ds_mesh.sample(mesh, snap_to_closest_point=True)`? It looks like your colors are categorical so averaging colors together is not what you want

**Reply by user27182** [reply]
There is also the option to set `categorical=True` to use nearest neighbor interpolation.
https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.sample.html#pyvista-datasetfilters-sample

**Reply by caleb-oconnor** [reply]
I had tried both of these approaches: 1) snap_to_clostest_point gave me the exact same results as ds_mesh.sample(mesh) and 2) categorical crashed giving this error "vtkProbeFilter.cxx:171    ERR| vtkCompositeDataProbeFilter (00000211CC236EA0): Source scalars have more than one component! Cannot categorize!"

**Comment by caleb-oconnor** [comment]
It seems that "decimate_pro" keeps the approximate color. I am not really familiar with the difference between decimate and decimate_pro but this is all I needed. Thanks everyone
![image](https://github.com/pyvista/pyvista/assets/93679057/600414bb-7fd9-4a94-892e-487fac3984ca)
